README.md
![td-io](/docs/photo.jpg)

td-io is a JVS I/O board designed to adapt older JAMMA-compatible cabinets to the newer JVS standard, used by modern arcade hardware such as the Naomi and exA. It features a number of independent systems to translate signals, drive cabinet hardware, and power JVS games.

# Features at a glance

* Chainable JVS I/O implementing a 2L12B panel and coin inputs
* JVS to JAMMA video converter compatible with all monitors
* Mono/stereo audio amplifier for driving JAMMA speakers
* JVS power connector providing 3.3V via an on-board converter
* High reliability, low heat generation and all solid state caps
* Upgradeable firmware

# Comparison with other JVS I/O options

|               | TD-IO   | Capcom I/O   | Sega 838   | Exa I/O |
| ------------- | ------- | ------------ | ---------- | ------- |
| Lag (frames)  | 0       | 1            | 0          | 0       |
| Chaining      | ✔ | ❌ | ✔ | ❌ |
| Video         | ✔ | ✔ | ✔ | ❌ |
| Audio amp     | ✔ | ✔ | ❌ | ❌ |
| 3.3V power    | ✔ | ✔ | ❌ | ❌ |
| Amplified stereo | ✔ | ❌ | ❌ | ❌ |

# JVS inputs

The JVS I/O bus connects to the board via USB-style A and B connectors. Daisy chaining is supported. The I/O presents two players, each with a single joystick/lever and six buttons. Joystick and button 1-4 inputs are mapped to the JAMMA edge connector, and buttons 4-6 are mapped to a CPS2-style kick harness connector. Buttons are polled synchronously, ensuring microsecond-level latency.

Service, test, and tilt buttons are also supported. JAMMA's single service button is mapped to JVS player 1's service button.

The board also presents one or two coin slots, toggleable by a DIP switch. The board also features two coin meter drivers and two coin lockout drivers. Coins are locked out when the respective coin count is set to its max value (16383).

# Video

The board maps JVS-level video (0.7Vp-p, AC or DC coupled) to JAMMA levels (0-3V, DC coupled). This is achieved via a DC restore step using the sync signal, followed by a fixed amplification. The DC coupled output allows maximum compatibility with monitors such as the Nanao MS8 and capture devices such as the Splitfire which require it.

In addition, a sync combiner combines negative H/V sync to negative composite sync.

No rate conversion is performed - the monitor must support the same scan rates as the JVS game.

An EDID EEPROM is also attached to the VGA port, to identify as a 31kHz monitor.

# Audio

TD-IO provides a class-D audio amplifier, which can drive both mono and stereo speakers in JAMMA cabinets. For mono cabinets, audio is simply provided via the JAMMA edge. For stereo support, the cabinet's speakers should be connected to the JST NH header provided on the I/O. The pinout of this header is directly compatible with the Egret 2, and adapters can be easily made for other cabinets, such as the Aero.

The stereo/mono switch selects whether to mixdown the L/R channels to mono. It should be set to mono when using a mono cabinet via the JAMMA edge, and stereo when using a cabinet with the stereo header.

The stereo support is not directly compatible with nonstandard Neo Geo style stereo over JAMMA, with a common ground. For these cabinets, either use the TD-IO in mono mode, or rewire the cabinet to connect directly to the stereo header. The left/right negative speaker outputs should not be tied together.

## Aero City stereo adapter

| JST NH H4P-SHF-AA | Molex MLX 50-84-2042 |
| --- | --- |
| 1   | 2   |
| 2   | 3   |
| 3   | 1   |
| 4   | 4   |

# Power

An 8-pin JVS power connector forwards 12V and 5V from the cabinet to the JVS board. In addition, an on-board power supply generates 3.3V at up to 7A from the cabinet's 5V rail for systems that require it, such as Naomi.

The cabinet's 5V power supply must still be adequate to power the 3.3V converter.

The IO's electronics are protected from overvoltage and overcurrent by an automatically resetting protector.

# Firmware Updates

The board features an on-board RP2040 microcontroller, whose firmware can be updated over USB for feature enhancements or compatibility fixes.

To update the firmware, turn off power to the cabinet (if attached). Then, attach a USB cable to the micro USB connector. Hold down the white button on the RP2040 while plugging the USB connector into a PC. The TD-IO will then appear as a flash drive. Copy the firmware file (ending in .uf2) onto the drive. The TD-IO will then disconnect and be ready for use.

# Analog Input

The board features an unpopulated expansion header exposing 3 3.3V level analog inputs, designed for potentiometers for analog joystick or racing games. This feature is experimental and not yet implemented in the firmware.

# Where to buy

The board is not currently for sale, as testing is not yet complete.

# License

CERN-OHL-W


CHANGELOG
1.3
---
Removed discrete EDID eeprom
Moved volume pot & fixed footprint
Added compatibility with TPA3137D2 amp
Removed logic power efuse
Added discrete TVS and efuse protection to coil drivers and JVS sense pin
Swapped 3.3V PSU to AOZ2262-11 based circuit
Changed button pullups to +5V

1.2
---
Changed negative voltage supply to MAX1720
Fixed RCA jack footprint
Added EDID emulation option
Fixed ADA4861 disable pin
Fixed PLIMIT pin on audio amp

1.1
---
Added DC restore circuit
Fixed audio amp filter wiring
Added +5V efuse
Reduced 3.3V feedback network impedance
Dropped 2.5V sense enable GPIO


1.0
---

First revision


firmware/ws2812.c
/**
 * Copyright (c) 2020 Raspberry Pi (Trading) Ltd.
 *
 * SPDX-License-Identifier: BSD-3-Clause
 */

#include <stdio.h>
#include <stdlib.h>

#include "pico/stdlib.h"
#include "hardware/pio.h"
#include "hardware/clocks.h"
#include "ws2812.pio.h"

static inline void put_pixel(uint32_t pixel_grb) {
    pio_sm_put_blocking(pio0, 0, pixel_grb << 8u);
}

static inline uint32_t urgb_u32(uint8_t r, uint8_t g, uint8_t b) {
    return
            ((uint32_t) (r) << 8) |
            ((uint32_t) (g) << 16) |
            (uint32_t) (b);
}

void pattern_snakes(uint len, uint t) {
    for (uint i = 0; i < len; ++i) {
        uint x = (i + (t >> 1)) % 64;
        if (x < 10)
            put_pixel(urgb_u32(0xff, 0, 0));
        else if (x >= 15 && x < 25)
            put_pixel(urgb_u32(0, 0xff, 0));
        else if (x >= 30 && x < 40)
            put_pixel(urgb_u32(0, 0, 0xff));
        else
            put_pixel(0);
    }
}

void pattern_snakes_green(uint len, uint t) {
  for (uint i = 0; i < len; ++i) {
    uint x = (i + (t >> 1)) % 64;
    if (x < 10)
      put_pixel(urgb_u32(0, 0x10, 0));
    else if (x >= 15 && x < 25)
      put_pixel(urgb_u32(0, 0x30, 0));
    else if (x >= 30 && x < 40)
      put_pixel(urgb_u32(0, 0xff, 0));
    else
      put_pixel(0);
  }
}

void pattern_snakes_green_fast(uint len, uint t) {
  for (uint i = 0; i < len; ++i) {
    uint x = (i + (t >> 0)) % 64;
    if (x < 10)
      put_pixel(urgb_u32(0, 0x10, 0));
    else if (x >= 15 && x < 25)
      put_pixel(urgb_u32(0, 0x30, 0));
    else if (x >= 30 && x < 40)
      put_pixel(urgb_u32(0, 0xff, 0));
    else
      put_pixel(0);
  }
}

void pattern_random(uint len, uint t) {
    if (t % 8)
        return;
    for (int i = 0; i < len; ++i)
        put_pixel(rand());
}

void pattern_sparkle(uint len, uint t) {
    if (t % 4)
        return;
    for (int i = 0; i < len; ++i)
        put_pixel(rand() % 8 ? 0 : 0xffffffff);
}

void pattern_sparkle_green(uint len, uint t) {
  if (t % 8)
    return;
  for (int i = 0; i < len; ++i) {
    uint8_t a = rand() % 16 ? 0 : 0xff;
    put_pixel(urgb_u32(0, 0xa, 0));
  }
}

void pattern_sparkle_yellow(uint len, uint t) {
  if (t % 4)
    return;
  for (int i = 0; i < len; ++i) {
    uint8_t a = rand() % 8 ? 0 : 0xff;
    put_pixel(urgb_u32(a, a, 0));
  }
}

void pattern_greys(uint len, uint t) {
    int max = 100; // let's not draw too much current!
    t %= max;
    for (int i = 0; i < len; ++i) {
        put_pixel(t * 0x10101);
        if (++t >= max) t = 0;
    }
}

typedef void (*pattern)(uint len, uint t);
const struct {
    pattern pat;
    const char *name;
} pattern_table[] = {
        {pattern_snakes,  "Snakes!"},
        {pattern_random,  "Random data"},
        {pattern_sparkle, "Sparkles"},
        {pattern_greys,   "Greys"},
};

const int PIN_TX = 27;

void init_ws2812(int pin) {
  // todo get free sm
  PIO pio = pio0;
  int sm = 0;
  uint offset = pio_add_program(pio, &ws2812_program);

  ws2812_program_init(pio, sm, offset, pin, 800000, false);
}

int run_ws2812() {
    //set_sys_clock_48();
    stdio_init_all();
    puts("WS2812 Smoke Test");

    // todo get free sm
    PIO pio = pio0;
    int sm = 0;
    uint offset = pio_add_program(pio, &ws2812_program);

    ws2812_program_init(pio, sm, offset, PIN_TX, 80000, false);

    int t = 0;
    while (1) {
        int pat = rand() % count_of(pattern_table);
        int dir = (rand() >> 30) & 1 ? 1 : -1;
        puts(pattern_table[pat].name);
        puts(dir == 1 ? "(forward)" : "(backward)");
        for (int i = 0; i < 1000; ++i) {
            pattern_table[pat].pat(150, t);
            sleep_ms(10);
            t += dir;
        }
    }
}


firmware/td-io.c
#include <stdio.h>
#include "pico/stdlib.h"
#include "hardware/adc.h"
#include <string.h>

const uint PIN_JVS_RE = 2;
const uint PIN_JVS_DE = 3;

const uint PIN_JVS_SENSE_2_5V = 14;
const uint PIN_JVS_SENSE_0V = 13;

const uint PIN_JVS_TERMINATION = 15;
const uint PIN_JVS_SENSE_IN_HIGH = 11;
const uint PIN_JVS_SENSE_IN_LOW = 12;

const uint PIN_SR_DATA = 20;
const uint PIN_SR_CLK = 18;
const uint PIN_SR_SH = 21;

const uint PIN_METER1 = 8;
const uint PIN_METER2 = 7;
const uint PIN_LOCKOUT1 = 10;
const uint PIN_LOCKOUT2 = 9;

const uint PIN_LED_ENUMERATED = PICO_DEFAULT_LED_PIN;
const uint PIN_DIP1 = 17;
const uint PIN_DIP2 = 16;

const uint16_t JVS_TERMINATION_THRESHOLD = (uint16_t)(3.75/2.0/3.3*4096);
const uint16_t JVS_0V_THRESHOLD = (uint16_t)(1.25/2.0/3.3*4096);

const uint8_t JVS_STATUS_GOOD = 1;
const uint8_t JVS_STATUS_UNKNOWN_COMMAND = 2;
const uint8_t JVS_STATUS_CHECKSUM_ERROR = 3;
const uint8_t JVS_STATUS_OVERFLOW = 4;

const uint8_t JVS_REPORT_GOOD = 1;
const uint8_t JVS_REPORT_PARAMETER_INVALID = 3;

const uint8_t JVS_MAX_LEN = 253; // minus two for status and checksum

// global state
uint8_t our_address = 0;
uint64_t last_process_coin = 0;
// signed for underflow checks
int16_t coin_count_p1 = 0;
int16_t coin_count_p2 = 0;
uint8_t prev_coin_p1 = 0;
uint8_t prev_coin_p2 = 0;

#define SR_P1_3 7
#define SR_P2_3 6
#define SR_P1_4 5
#define SR_P2_4 4
#define SR_P1_5 3
#define SR_P2_5 2
#define SR_P1_6 1
#define SR_P2_6 0
#define SR_P1_LEFT 15
#define SR_P2_LEFT 14
#define SR_P1_RIGHT 13
#define SR_P2_RIGHT 12
#define SR_P1_1 11
#define SR_P2_1 10
#define SR_P1_2 9
#define SR_P2_2 8
#define SR_C1 23
#define SR_C2 22
#define SR_P1_START 21
#define SR_P2_START 20
#define SR_P1_UP 19
#define SR_P2_UP 18
#define SR_P1_DOWN 17
#define SR_P2_DOWN 16
#define SR_SERVICE 31
#define SR_TEST 30
#define SR_TILT 29

const uint8_t JVS_COMM_VER = 0x10;
const char id_str[] = "TD;TD-IO;v1.1;https://github.com/tdaede/td-io";

const uint8_t input_desc_1coin[] = {
    0x01, 2, 12, 0,
    0x02, 1, 0, 0,
    0x00
};

const uint8_t input_desc_2coin[] = {
    0x01, 2, 12, 0,
    0x02, 2, 0, 0,
    0x00
};

void jvs_sync() {
    uart_putc(uart0, 0xe0);
}

uint8_t jvs_getc() {
    uint8_t c = uart_getc(uart0);
    if (c == 0xd0) {
        c = uart_getc(uart0);
        return c + 1;
    } else {
        return c;
    }
}

void jvs_putc(uint8_t c) {
    if (c == 0xe0) {
        uart_putc(uart0, 0xd0);
        uart_putc(uart0, 0xdf);
    } else if (c == 0xd0) {
        uart_putc(uart0, 0xd0);
        uart_putc(uart0, 0xcf);
    } else {
        uart_putc(uart0, c);
    }
}

void start_transmit() {
    gpio_put(PIN_JVS_RE, 1); // disable receive
    busy_wait_us_32(200);    // 200us min for exa compat
    gpio_put(PIN_JVS_DE, 1); // enable transmitter
}

void stop_transmit() {
    uart_tx_wait_blocking(uart0);
    gpio_put(PIN_JVS_DE, 0); // disable transmitter
    gpio_put(PIN_JVS_RE, 0); // enable receiver
}

void send_message(uint8_t status, uint8_t* m, uint8_t msg_len) {
    start_transmit();
    jvs_sync();
    uint8_t checksum = 0;
    jvs_putc(0);
    checksum += 0;
    jvs_putc(msg_len + 2);
    checksum += msg_len + 2;
    jvs_putc(status);
    checksum += status;
    for (int i = 0; i < msg_len; i++) {
        jvs_putc(m[i]);
        checksum += m[i];
    }
    jvs_putc(checksum);
    stop_transmit();
}

uint32_t read_switches() {
    uint32_t r;
    gpio_put(PIN_SR_SH, 1);
    busy_wait_us(1);
    for (int i = 0; i < 32; i++) {
        r >>= 1;
        r |= (gpio_get(PIN_SR_DATA) ? 1 : 0) << 31;
        gpio_put(PIN_SR_CLK, 1);
        busy_wait_us(1);
        gpio_put(PIN_SR_CLK, 0);
        busy_wait_us(1);
    }
    gpio_put(PIN_SR_SH, 0);
    return ~r;
}

void update_termination() {
    if (gpio_get(PIN_JVS_SENSE_IN_HIGH)) {
        gpio_put(PIN_JVS_TERMINATION, 0);
    } else {
        gpio_put(PIN_JVS_TERMINATION, 1);
    }
}

// call periodically with switch read to check level changes of coin input
void process_coin(uint32_t switches) {
    if ((switches >> SR_C1 & 1) && !prev_coin_p1) {
        coin_count_p1++;
        if (coin_count_p1 > 16383) {
            coin_count_p1 = 16383;
        }
    }
    if ((switches >> SR_C2 & 1) && !prev_coin_p2) {
        coin_count_p2++;
        if (coin_count_p2 > 16383) {
            coin_count_p2 = 16383;
        }
    }
    prev_coin_p1 = switches >> SR_C1 & 1;
    prev_coin_p2 = switches >> SR_C2 & 1;
    gpio_put(PIN_METER1, prev_coin_p1);
    gpio_put(PIN_METER2, prev_coin_p2);
    gpio_put(PIN_LOCKOUT1, coin_count_p1 >= 16383);
    gpio_put(PIN_LOCKOUT2, coin_count_p2 >= 16383);
}

int main() {
    stdio_init_all();
    uart_init(uart0, 115200);
    uart_set_translate_crlf(uart0, false);
    gpio_set_function(0, GPIO_FUNC_UART);
    gpio_set_function(1, GPIO_FUNC_UART);
    gpio_init(PIN_JVS_RE);
    gpio_put(PIN_JVS_RE, 0); // enable receiver
    gpio_set_dir(PIN_JVS_RE, GPIO_OUT);
    gpio_init(PIN_JVS_DE);
    gpio_put(PIN_JVS_DE, 0); //disable transmitter
    gpio_set_dir(PIN_JVS_DE, GPIO_OUT);
    gpio_init(PIN_JVS_SENSE_2_5V);
    gpio_put(PIN_JVS_SENSE_2_5V, 1); // always appear present
    gpio_set_dir(PIN_JVS_SENSE_2_5V, GPIO_OUT);
    gpio_init(PIN_JVS_SENSE_0V);
    gpio_put(PIN_JVS_SENSE_0V, 0);
    gpio_set_dir(PIN_JVS_SENSE_0V, GPIO_OUT);
    gpio_init(PIN_JVS_TERMINATION);
    gpio_put(PIN_JVS_TERMINATION, 1); // disable termination by default
    gpio_set_dir(PIN_JVS_TERMINATION, GPIO_OUT);
    gpio_init(PIN_JVS_SENSE_IN_LOW);
    gpio_set_dir(PIN_JVS_SENSE_IN_LOW, GPIO_IN);
    gpio_pull_up(PIN_JVS_SENSE_IN_LOW);
    gpio_init(PIN_JVS_SENSE_IN_HIGH);
    gpio_set_dir(PIN_JVS_SENSE_IN_HIGH, GPIO_IN);
    gpio_pull_up(PIN_JVS_SENSE_IN_HIGH);

    // sr
    gpio_init(PIN_SR_DATA);
    gpio_set_dir(PIN_SR_DATA, GPIO_IN);
    gpio_init(PIN_SR_CLK);
    gpio_put(PIN_SR_CLK, 0);
    gpio_set_dir(PIN_SR_CLK, GPIO_OUT);
    gpio_init(PIN_SR_SH);
    gpio_put(PIN_SR_SH, 0);
    gpio_set_dir(PIN_SR_SH, GPIO_OUT);

    // coin/lockout drivers
    gpio_init(PIN_METER1);
    gpio_put(PIN_METER1, 0);
    gpio_set_dir(PIN_METER1, GPIO_OUT);
    gpio_init(PIN_METER2);
    gpio_put(PIN_METER2, 0);
    gpio_set_dir(PIN_METER2, GPIO_OUT);
    gpio_init(PIN_LOCKOUT1);
    gpio_put(PIN_LOCKOUT1, 0);
    gpio_set_dir(PIN_LOCKOUT1, GPIO_OUT);
    gpio_init(PIN_LOCKOUT2);
    gpio_put(PIN_LOCKOUT2, 0);
    gpio_set_dir(PIN_LOCKOUT2, GPIO_OUT);

    // ui
    gpio_init(PIN_LED_ENUMERATED);
    gpio_put(PIN_LED_ENUMERATED, 0);
    gpio_set_dir(PIN_LED_ENUMERATED, GPIO_OUT);
    gpio_init(PIN_DIP1);
    gpio_set_dir(PIN_DIP1, GPIO_IN);
    gpio_pull_up(PIN_DIP1);

    update_termination();

    while (true) {
        const uint64_t now = time_us_64();
        if ((now - last_process_coin) > 20000) {
            uint32_t switches = read_switches();
            process_coin(switches);
            last_process_coin = now;
        }

        if (uart_is_readable(uart0) && uart_getc(uart0) == 0xe0) {
            uint8_t our_checksum = 0;
            uint8_t node_num = jvs_getc();
            if (!((node_num == 0xff) || ((our_address == node_num) && (our_address != 0)))) {
                continue;
            }
            our_checksum += node_num;
            uint8_t length = jvs_getc();
            our_checksum += length;
            uint8_t msg_length = length - 1;
            uint8_t message[256];
            for (int i = 0; i < msg_length; i++) {
                uint8_t c = jvs_getc();
                our_checksum += c;
                message[i] = c;
            }
            uint8_t msg_send[256*2]; // a few bytes to spare for easier overflow checking
            int o = 0;
            uint8_t status = JVS_STATUS_GOOD;
            uint8_t their_checksum = jvs_getc();
            if (our_checksum == their_checksum) {
                int i = 0;
                while (i < msg_length) {
                    if ((msg_length - i) >= 2 && message[i] == 0xf0 && message[i+1] == 0xd9) {
                        printf("N: %02x Reset\n", node_num);
                        our_address = 0;
                        gpio_put(PIN_JVS_SENSE_0V, 0);
                        gpio_put(PIN_LED_ENUMERATED, 0);
                        i += 2;
                    } else if ((msg_length - i) >= 2 && message[i] == 0xf1) {
                        uint8_t node_id = message[i+1];
                        i += 2;
                        printf("Assign node id N: %02x\n", node_id);
                        if ((our_address == 0)
                            && (gpio_get(PIN_JVS_SENSE_IN_HIGH)
                                || !gpio_get(PIN_JVS_SENSE_IN_LOW))) {
                            printf("Assigning our address\n");
                            our_address = node_id;
                            msg_send[o] = JVS_REPORT_GOOD;
                            o++;
                            gpio_put(PIN_JVS_SENSE_0V, 1);
                            gpio_put(PIN_LED_ENUMERATED, 1);
                        } else {
                            printf("We are not currently last in the chain, skipping assignment\n");
                        }
                    } else if ((msg_length - i) >= 1 && message[i] == 0x10) {
                        i++;
                        printf("Got ID code request\n");
                        msg_send[o] = JVS_REPORT_GOOD;
                        o++;
                        strcpy((char*)&msg_send[o], id_str);
                        o += strlen(id_str) + 1;
                    } else if ((msg_length - i) >= 1 && message[i] == 0x11) {
                        i++;
                        printf("Got revision request\n");
                        msg_send[o] = JVS_REPORT_GOOD;
                        o++;
                        msg_send[o] = 0x13;
                        o++;
                    } else if ((msg_length - i) >= 1 && message[i] == 0x12) {
                        i++;
                        printf("Got video revision request\n");
                        msg_send[o] = JVS_REPORT_GOOD;
                        o++;
                        msg_send[o] = 0x30;
                        o++;
                    } else if ((msg_length - i) >= 1 && message[i] == 0x13) {
                        i++;
                        printf("Got comm revision request\n");
                        msg_send[o] = JVS_REPORT_GOOD;
                        o++;
                        msg_send[o] = JVS_COMM_VER;
                        o++;
                    } else if ((msg_length - i) >= 1 && message[i] == 0x14) {
                        i++;
                        printf("Got input descriptor request\n");
                        msg_send[o] = JVS_REPORT_GOOD;
                        o++;
                        if (!gpio_get(PIN_DIP1)) {
                            memcpy(&msg_send[o], input_desc_1coin, sizeof(input_desc_1coin));
                            o += sizeof(input_desc_1coin);
                        } else {
                            memcpy(&msg_send[o], input_desc_2coin, sizeof(input_desc_2coin));                        o += sizeof(input_desc_2coin);
                        }
                    } else if ((msg_length - i) >= 1 && message[i] == 0x15) {
                        i++;
                        printf("Got main board ID: ");
                        while (i < msg_length) {
                            char c = message[i];
                            i++;
                            if (c == 0) break;
                            putchar(message[i]);
                        }
                        printf("\n");
                        msg_send[o] = JVS_REPORT_GOOD;
                        o++;
                    } else if ((msg_length - i) >= 3 && message[i] == 0x20) {
                        int num_players = message[i+1];
                        int bytes_per_player = message[i+2];
                        i += 3;
                        if (o + num_players*bytes_per_player >= JVS_MAX_LEN) {
                            printf("JVS response overflow!\n");
                            status = JVS_STATUS_OVERFLOW;
                            break;
                        }
                        msg_send[o] = JVS_REPORT_GOOD;
                        o++;
                        uint32_t switches = read_switches();
                        if ((now - last_process_coin) > 12000) {
                            process_coin(switches);
                            last_process_coin = now;
                        }
                        msg_send[o] = ((switches >> SR_TEST) & 1) << 7
                            | ((switches >> SR_TILT) & 1) << 7;
                        o++;
                        //printf("Got switch request for %02x players\n", num_players);
                        for (int player = 0; player < num_players; player++) {
                            for (int byte = 0; byte < bytes_per_player; byte++) {
                                uint8_t b = 0;
                                if ((player == 0) && (byte == 0)) {
                                    b = ((switches >> SR_P1_2) & 1) << 0
                                        | ((switches >> SR_P1_1) & 1) << 1
                                        | ((switches >> SR_P1_RIGHT) & 1) << 2
                                        | ((switches >> SR_P1_LEFT) & 1) << 3
                                        | ((switches >> SR_P1_DOWN) & 1) << 4
                                        | ((switches >> SR_P1_UP) & 1) << 5
                                        | ((switches >> SR_SERVICE) & 1) << 6
                                        | ((switches >> SR_P1_START) & 1) << 7;
                                } else if ((player == 0) && (byte == 1)) {
                                    b = ((switches >> SR_P1_6) & 1) << 4
                                        | ((switches >> SR_P1_5) & 1) << 5
                                        | ((switches >> SR_P1_4) & 1) << 6
                                        | ((switches >> SR_P1_3) & 1) << 7;
                                } else if ((player == 1) && (byte == 0)) {
                                    b = ((switches >> SR_P2_2) & 1) << 0
                                        | ((switches >> SR_P2_1) & 1) << 1
                                        | ((switches >> SR_P2_RIGHT) & 1) << 2
                                        | ((switches >> SR_P2_LEFT) & 1) << 3
                                        | ((switches >> SR_P2_DOWN) & 1) << 4
                                        | ((switches >> SR_P2_UP) & 1) << 5
                                        | ((switches >> SR_P2_START) & 1) << 7;
                                } else if ((player == 1) && (byte == 1)) {
                                    b = ((switches >> SR_P2_6) & 1) << 4
                                        | ((switches >> SR_P2_5) & 1) << 5
                                        | ((switches >> SR_P2_4) & 1) << 6
                                        | ((switches >> SR_P2_3) & 1) << 7;
                                }
                                msg_send[o] = b;
                                o++;
                            }
                        }
                    } else if ((msg_length - i) >= 2 && message[i] == 0x21) {
                        int slots = message[i+1];
                        i += 2;
                        if (o + slots*2 >= JVS_MAX_LEN) {
                            printf("JVS response overflow!\n");
                            status = JVS_STATUS_OVERFLOW;
                            break;
                        }
                        msg_send[o] = JVS_REPORT_GOOD;
                        o++;
                        //printf("Got coin slot request for %02x slots\n", slots);
                        for (int slot = 0; slot < slots; slot++) {
                            if (slot == 0) {
                                msg_send[o] = coin_count_p1 >> 8;
                                msg_send[o+1] = coin_count_p1 & 0xFF;
                            } else if (slot == 1) {
                                msg_send[o] = coin_count_p2 >> 8;
                                msg_send[o+1] = coin_count_p2 & 0xFF;
                            } else {
                                msg_send[o] = 0x80;
                                msg_send[o+1] = 0x00;
                            }
                            o += 2;
                        }
                    } else if ((msg_length - i) >= 4 && message[i] == 0x30) {
                        uint8_t slot = message[i+1];
                        uint16_t amount = (message[i+2] << 8) + message[i+3];
                        i += 4;
                        printf("Decrement coin counter %d by %d\n", (int)slot, (int)amount);
                        if (amount > 16383) {
                            amount = 16383;
                            printf("Capped to 16383\n");
                        }
                        if (slot == 1) {
                            msg_send[o] = JVS_REPORT_GOOD;
                            coin_count_p1 -= amount;
                            if (coin_count_p1 < 0) {
                                coin_count_p1 = 0;
                            }
                        } else if (slot == 2) {
                            msg_send[o] = JVS_REPORT_GOOD;
                            coin_count_p2 -= amount;
                            if (coin_count_p2 < 0) {
                                coin_count_p2 = 0;
                            }
                        } else {
                            printf("Invalid coin counter slot\n");
                            msg_send[o] = JVS_REPORT_PARAMETER_INVALID;
                        }
                        o++;
                    } else if ((msg_length - i) >= 4 && message[i] == 0x35) {
                        uint8_t slot = message[i+1];
                        uint16_t amount = (message[i+2] << 8) + message[i+3];
                        i += 4;
                        printf("Increment coin counter %d by %d\n", (int)slot, (int)amount);
                        if (amount > 16383) {
                            amount = 16383;
                            printf("Capped to 16383\n");
                        }
                        if (slot == 1) {
                            msg_send[o] = JVS_REPORT_GOOD;
                            coin_count_p1 += amount;
                            if (coin_count_p1 > 16383) {
                                coin_count_p1 = 16383;
                            }
                        } else if (slot == 2) {
                            msg_send[o] = JVS_REPORT_GOOD;
                            coin_count_p2 += amount;
                            if (coin_count_p2 > 16383) {
                                coin_count_p2 = 16383;
                            }
                        } else {
                            printf("Invalid coin counter slot\n");
                            msg_send[o] = JVS_REPORT_PARAMETER_INVALID;
                        }
                        o++;
                    } else {
                        printf("Unsupported message: N: %02x L: %02x M: ", node_num, msg_length);
                        for (int j = 0; j < msg_length; j++) {
                            printf("%02x", message[j]);
                        }
                        printf("\n");
                        status = JVS_STATUS_UNKNOWN_COMMAND;
                        break;
                    }
                    if (o >= JVS_MAX_LEN) {
                        printf("JVS response overflow!\n");
                        status = JVS_STATUS_OVERFLOW;
                        break;
                    }
                }
                fflush(stdout);
                if ((o > 0) || (status != JVS_STATUS_GOOD)) {
                    send_message(status, msg_send, o);
                }
            } else {
                printf("Checksum mismatch: theirs: %02x ours: %02x\n", their_checksum, our_checksum);
                status = JVS_STATUS_CHECKSUM_ERROR;
                send_message(status, msg_send, 0);
            }
        } else {
            // note because of disabling the rx, we always read a 00 byte after sending here
            update_termination(); // convenient time to read adc
            //printf("Saw non-sync code %02x\n", sync);
        }
    }
    return 0;
}
